[Sema] NFC: More TVO_CanBindToInOut removal

davezarzycki opened the pull request:
This is a followup to #15580 and #15574, where @CodaFi pointed out that @slavapestov added TVO_CanBindToInOut to the callers of createTypeVariable out of an "abundance of caution". This pull request removes all-but-one of the remaining uses of TVO_CanBindToInOut.

For the record and unlike #15580, there is a small cost to diagnostic quality in this pull request, but if the goal is to remove `InOutType`, then I imagine that some amount of diagnostic churn is expected and tolerable.

davezarzycki committed:
[Sema] NFC: More TVO_CanBindToInOut removal

If we're willing to tolerate less specific diagnostics, then we can
remove more instances of TVO_CanBindToInOut. After this change, there is
only one client of TVO_CanBindToInOut.

davezarzycki commented:
@swift-ci please test

rudkx reviewed (commented):
I'll defer to @xedin on whether the diagnostic regressions are acceptable or if he has thoughts on how we can return to more reasonable diagnostics here.

xedin reviewed (approved):
Overall looks good to me, but it would be great if diagnostics are dealt with at the same time.

slavapestov commented:
@davezarzycki My most recent idea was to keep InOutExpr, but diagnose bad placement in preCheckExpression(). Then the solver will only see InOutExpr appear in function call argument position, and any mismatches there are already handled by CSDiag. This eliminates the need to diagnose InOutExpr in "weird" places and should allow us to consolidate various diagnostic paths around inout. If you'd like to try tackling this, I would be very happy :)

davezarzycki commented:
Hi @xedin – I have the parser changes roughly working. What diagnostic message would you like when `tok::amp_prefix` is found in an invalid position?

xedin commented:
@davezarzycki I think you can use `extraneous_address_of`, that would be good enough.

xedin commented:
But you'll have to add that to `DiagnosticsParser.def` as well.

davezarzycki commented:
What about binary operators? Should we warn or just continue to tolerate `&x == ptr`, etc.

xedin commented:
If we wanted to change that, I believe it would require swift-evolution pitch/proposal.

slavapestov commented:
If you do the diagnosis in pre-checking instead of parsing, you can handle operators without breaking compatibility since they get folded to ApplyExprs in that phase.

davezarzycki committed:
[Parse] Move tok::amp_prefix closer to parameter list logic

The amp_prefix token is currently tolerated in any unary expression
context and then diagnosed later by Sema. This patch changes parsing to
only accept tok::amp_prefix in its allowed position: parameter lists.

This also fixes two "compiler crasher" tests.

davezarzycki commented:
The parser change passes validation on my Linux machine.

@swift-ci please smoke test

davezarzycki commented:
@swift-ci please smoke test

davezarzycki commented:
@swift-ci please smoke test osx

davezarzycki commented:
@swift-ci please smoke test macos

xedin commented on test/Parse/pointer_conversion.swift.gyb:
```diff
@@ -62,7 +62,9 @@ func mutablePointerArguments(_ p: UnsafeMutablePointer<Int>,
   // We don't allow these conversions outside of function arguments.
+  //expected-error @+1 {{use of extraneous '&'}}
   var x: UnsafeMutablePointer<Int> = &i // expected-error{{cannot pass immutable value as inout argument: 'i' is immutable}}
+  //expected-error @+1 {{use of extraneous '&'}}
```
Why are there multiple diagnostics now which refer to the same thing? It seems like `preCheckExpression` might be a better place to diagnose this problem to avoid double diagnosing expressions...

davezarzycki replied:
Well, this is technically two errors. One by parsing and one by the type checker. The third error is for the subsequent statement. I can certainly tweak parsing to not emit `InOutExpr` if it knows that it is will be diagnosed again later. I can see the argument for this on the grounds that if the `&` is truly extraneous, then it shouldn't generate trickle-down errors in Sema.

xedin replied:
I think there is no reason to type-check in this situation because we'd know that expression is invalid if diagnosing was done by `preCheckExpression`, multiple errors like this could be really confusing...

davezarzycki replied:
I'll update the patch to hopefully improve diagnostic output quality.

xedin replied:
Thank you!

davezarzycki commented:
Hi @xedin – I made the parsing changes that you requested. This invalided two of the "compiler_crasher" regression tests, so I deleted them. Was that the right solution?

@swift-ci please smoke test

davezarzycki commented:
@swift-ci please smoke test macos

xedin commented:
@davezarzycki That's awesome! There is actually a script to resolve them instead of deleting, we'd really want keep them around to make sure compiler doesn't regress in the future - `./utils/resolve-crashes.py` pipe the output of your build with failing crashers to the script and it would do the rest (move them into *_fixed directory and remove `REQUIRES: asserts` I believe).

davezarzycki commented:
@swift-ci please test

davezarzycki commented:
@swift-ci please test

davezarzycki commented:
@swift-ci please test macos

davezarzycki commented:
I might need to create a new pull request. CI seems cranky.

xedin commented:
Most likely just saturated because too many people are test their PRs during the day :)

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

davezarzycki commented:
@swift-ci please test source compatibility

davezarzycki commented:
The source compatibility failures look unrelated. Lots of SIL deserialization failures with `$SSo20NSCalendarIdentifieras8HashableSCsACP9hashValueSivgTW`, which doesn't involve `inout` as far as I can tell.

@xedin – Do you agree? Is this safe to merge?

xedin commented:
Change which caused this has been reverted, but let's re-run to see if these failures could have hidden anything related to this PR.

xedin commented:
@swift-ci please test source compatibility

xedin commented:
Parser changes look good to me as well, I think once source compatibility is finished as green, this changes are good to go.
